### PR TITLE
feat: add extractLsByte(s) API for memory.

### PR DESCRIPTION
**Pull request by bollu**
This allows us to eventually write functions in the memory model that work byte-at-a-time.
This PR is peeled from https://github.com/leanprover/LNSym/pull/56.

### CODE REVIEW

File: Arm/BitVec.lean [hunk 1]
```diff
@@ -813,4 +813,84 @@ macro_rules
...
+    (decide (i ≤ 7) && val.getLsb (n * 8 + i)) := by
+  simp [extractLsByte]
+  rw [Nat.succ_mul]
+  simp only [Nat.add_one_sub_one]
+  simp [Nat.add_sub_cancel_left]
+
+
+/-! ### Least Significant Byte range -/
+
+/-- grab 'n' least significant bytes, starting from index 'base'.
```
Comment (shigoel) [inline Arm/BitVec.lean:880]:
```suggestion
/-- Get `n` least significant bytes of `val`, starting from index `base`.
```

File: Arm/BitVec.lean [hunk 2]
```diff
@@ -813,4 +813,84 @@ macro_rules
...
+-/
+def extractLsBytes (val : BitVec w) (base : Nat) (n : Nat) : BitVec (n * 8) :=
+  match h : n with
+  | 0 => 0#0
+  | x + 1 => val.extractLsb (base * 8 + n * 8 - 1) (base * 8) |>.cast (by omega)
+
+@[simp]
+theorem getLsb_extractLsBytes (val : BitVec w) (base : Nat) (n : Nat) (i : Nat) :
+    (BitVec.extractLsBytes val base n).getLsb i =
+      (decide (0 < n) && (decide (i ≤ base * 8 + (n) * 8 - 1 - base * 8) && val.getLsb (base * 8 + i))) := by
```
Comment (shigoel) [inline Arm/BitVec.lean:891]: Quibble: let's try to wrap long lines at column 80.

File: Arm/BitVec.lean [hunk 3]
```diff
@@ -813,4 +813,84 @@ macro_rules
...
+  val.extractLsb ((base + 1)*8 - 1) (base * 8) |>.cast (by omega)
+
+theorem extractLsByte_def (val : BitVec w₁) (n : Nat) :
+    val.extractLsByte n = (val.extractLsb ((n + 1)*8 - 1) (n * 8) |>.cast (by omega)) := rfl
+
+@[simp]
+theorem getLsb_extractLsByte (val : BitVec w₁) :
+    ((BitVec.extractLsByte val n).getLsb i) =
+    (decide (i ≤ 7) && val.getLsb (n * 8 + i)) := by
+  simp [extractLsByte]
```
Comment (shigoel) [inline Arm/BitVec.lean:872]: Stick to `simp only` in LNSym as far as possible.

File: Arm/BitVec.lean (hunk 2 above)
Comment (bollu) [reply Arm/BitVec.lean:891]: [Mathlib allows column length 100](https://github.com/leanprover-community/mathlib4/blob/055d766c68bd10e70d7adf8fe87f44add3d806fa/Mathlib/Tactic/Linter/Lint.lean#L259C9-L260). Happy to switch to column length 80 if desired.

**Comment by bollu** [comment]
@shigoel all comments addressed.

File: Arm/BitVec.lean (hunk 3 above)
Comment (bollu) [reply Arm/BitVec.lean]: Fixed.

File: Arm/BitVec.lean (hunk 2 above)
Comment (shigoel) [reply Arm/BitVec.lean]: Yes, please. Thanks!

**Review by shigoel — APPROVED** [review]

**Pull request merged**